docs/CppInteroperability/Roadmap.md

```diff
@@ -0,0 +1,141 @@
+
+As per our goal’s specification, this method of importing reference types allows C++ interoperability to have a clear, native mapping for a common C++ API pattern that builds a more general purpose solution off of the pre-established reference type bridging from Objective-C. Additionally, Swift preserves the same safety properties as C++ while providing an even safer option for the common pattern where reference types have a retain and release operation.
+
+### Value types
+
+As mentioned above, swift types fall into two categories: value types and reference types. Value types are relatively simple in Swift, C, and Objective-C, however, the same is not true for C++ where value types may have complex lifetime operations that are used to express non-native concepts (to Swift) such as value types that own memory. This case of owned value types is specifically difficult to map on to Swift as it is novel to the language and does not always work well with the existing model for Swift value types that are trivially copyable.  
+
+#### Trivial types
+
+This roadmap will refer to trivial value types that do not hold pointers as “trivial types.” These types include primitive types such as integers and types which are composed of other trivial types. Trivial types are “owned” types that provide trivial lifetime operations: a copy is a copy of their bits and a destroy is a no-op. Trivial types have roughly the same mapping throughout Swift, C, Objective-C, and C++ making them trivial to import. Trivial types, their instances, methods on trivial types, and other APIs that use trivial types are generally considered to be safe and usable.
+
+**Pointer types**
+
+This roadmap will refer to trivial value types that hold pointers as “pointer types.” These types include pointers themselves and types which are composed of any pointer types (potentially including other types as well). The pointers held by pointer types refer to memory that is *not owned* by the pointer type (making pointer types a “view” or “projection” into memory). While pointer types are very similar to trivial types with respect to their lifetime operations and the fact that they map similarly in these four language, they differ in the fact that while they themselves are not inherently unsafe, they may be used in unsafe APIs (discussed later).
```

Comment:
"The pointers held by pointer types refer to memory that is *not owned* by the pointer type (making pointer types a “view” or “projection” into memory)." <- Do I understand this correctly that these only exclude pointers to Owned types as defined below? (If so, maybe state this explicitly. If not, need to explain more precisely what you mean.) If owned types require an annotation, how do we differentiate a case of a safe pointer type with a pointer type pointing to owned memory, if the user forgets to annotate the owned type? Or do we assume that the annotations user provides recorrect in our safety model?

Reply:
Yes, these exclude owned types. I will state that explicitly. 

Here's where copy constructors come into play (and my question "can anyone think of an owned type that holds a pointer and has a trivial (implicit, compiler provided) copy constructor"). Assuming the Swift compiler can correctly differentiate between "owned types" and "pointer types" using special members, I think we are OK here.

Reply:
I guess any type that has a "pointer" and doesn't have a explicit copy constructor is a pointer type.  
> can anyone think of an owned type that holds a pointer and has a trivial (implicit, compiler provided) copy constructor")
Is `S` an Owned type or Pointer type?  If this is a Pointer type (which I suppose), is this code allowed and is it less safe in Swift? 

```
var v = vector(1) 
let start = v.begin() 
doSomething(start) 
fixLifetime(v)
```

```
struct Record {
   int f1, f2 ... ; 
   // huge record consisting of ints/bools/floats and other structs composed of that. 
}; 

struct S { 
   int &unownedReference; 
   Record record; 

   Record& getRecord() { return record; }
}
```

Reply:
Could I simplify your example to make it:
```
struct S {
  int x;
  int &getRecord() { return x; }
};
```

In this case, `getRecord` is projecting out an inner pointer from a trivial type. I suppose there are two courses of action here:
1. Treat trivial types in the same was as other owned types (with respect to banning projections)
2. Deem this kind of projection sufficiently safe

I tend towards the latter as these kinds of projections are not foreign to Swift (or C for that matter). The counter argument is 1) these kinds of projections are much more common in C++ where references are used for performance and 2) methods make this easier to get wrong. Maybe the solution here is to always borrow self and transform these references into getters (that return values) and setters?

Reply:
Thanks for clarifying. 
1. So IIUC, whatever behavior such methods have on trivial types will have will carry over to pointer types as they are just trivial types with some pointers  
2. **A silenceable compiler warning could be an option for these unsafe cases:**  In C interop, there are cases where if an implicit conversion from `String` to `char *` can result in a short-lived pointer, the Swift compiler presents a warning (this works through the @_ephemeral attributes). I am wondering if in these cases can have a similar compiler warning. But I think in that case there should be a syntactic way to silence the warning, either through a different safer construct (I suspect that doesn't exist here in general, unlike in the String -> char * conversion where a scoped call like `withCString` would be a safer option), or through some other syntactic decoration (adding additional parantheses etc.). This would enable the programmer to express that they are confident of the correctness, because may be `getRecord` is not returning a pointer to the storage but something that would be alive for much longer.

Reply:
In fact, such a silenceable compiler a warning could also be an option in general for allowing unsafe projection on types with storage.

Reply:
@ravikandhadai your point here is that you think we should introduce a compiler warning for these kinds of unsafe APIs that fall between "projections on owned types" and "safe methods on trivial types"? 

As I've said before, I think a "safe mode" for the clang importer, or some extra safety warnings like this could be a good idea, but we need to detach that "feature" from C++ interop. It should be applied to C and Objective-C interop as well and, while it's related to importing C++ APIs, it's not necessary for our user model (not something that should block interop design).

Reply:
Actually, I am just saying it is an alternative to allow everything and have a slienceable warning where you already ban APIs from automatically importing.

Reply:
So not ban anything but have warnings instead, that can be silenced. If so, it can be more aggressive.

Reply:
I think the difference here is that projections aren't disallowed because they are too unsafe, they are disallowed because we simply don't know how to import them. The unsafety we are discussing is a side-effect of a naive import of these projects. Warnings indicate that these APIs are just "too unsafe" when in reality we don't have the language (ownership) features to be able to correctly represent them. As Anna suggested, I will elaborate on how that could be done once the required ownership features have landed.

Reply:
> I think the difference here is that projections aren't disallowed because they are too unsafe, they are disallowed because we simply don't know how to import them. 

If we can precisely tell in both sound and complete way when is a method projecting and when it is not, that's a different story. But those are semantic properties and by rice theorem we can only either conservatively over-approximate or be unsound and incomplete. Here are we are choosing to be unsound and incomplete (which is always a practical choice). Then users would always have methods/code snippets that are eliminated only because they aren't conforming to the model, not because they are semantically unsafe. So an alternative would be to turn this "error", which completely disallows based on an model, into a warning. That puts the final decision back into the hands of the user. The warning can detail what is the problem, and if users confirm that is not a problem, they can suppress it and have it imported.  The more the model becomes restrictive the more cases where having it as a warning would make sense.

Reply:
The knobs (read annotations) that is proposed here are at the type level (not at method level, though the problem is at the method level). And those knobs all require the library author to make those changes to permit them. Which is a blanket approval that allows all methods on those types to be used in all context. So a warning on the client side, could be context specific and a fine granularity like a method.

Reply:
At a more practical, the user model can be thought of a providing a compiler "error" in cases that are disallowed. The alternative would be to make it a warning that can be suppressed. 

In cases where the user model is restrictive or that is good amount of uncertainty on its restrictiveness, warnings are a reasonable alternative.

Reply:
Warnings can also let you be little bit more relaxed on where to warn, since there is a way to suppress. So it allows warning on more code snippets (e.g. the user model doesn't have to so carefully tune the safety/restrictiveness). E.g. it is possible to warn on a method that returns a `std::vector<StringRef>` and call it a projection and not worry about disallowing a huge class of methods unusable.

Reply:
I think this is a great idea that we've often discussed but I neglected to capture in this document: we should have an annotation that can go on both types and individual methods/functions which allows them to be imported regardless of the rules we define here. The inverse already exists with the swift unavailable attribute. 

I don't see why this needs to be tied to a warning/error, though.

Reply:
Method-level granularity by itself is independent of error/warning. 

It does raise an interesting question of whether those suppressions need to be context specific e.g. if it depends on template instantiations and not possible to define it on the method definition. 

> Warnings vs error is a separate question

The factors that would determine this is. 
1. If this suppression has to be client-context specific
2. How much false posistive/negative this user model is going to have. 

The second questions is probably really hard to answer.